File: experiments/runs/2026-09-20_10-40-56_claim-office-java-example-mapping_exact-ptdd-v1.1-refactor-subagent-cc_opus-5-no-thinking/src/main/java/PayoutCap.java
import java.math.BigDecimal;
import java.util.List;

/**
 * The ceiling the MHPCO puts on what one policy ever pays out, and what is
 * left of it.
 *
 * <p>The MHPCO pays out at most twice the insurance sum over the whole life of
 * a policy. The cap is measured against the unmodified insurance sum: the
 * premium modifiers and the building block offer bear on what the customer
 * pays, never on what the MHPCO is willing to pay back. Once the cap is
 * exhausted a claim is reduced to whatever is left of it rather than refused.
 *
 * <p>Changes whenever the MHPCO revises how far it will go for a single
 * policy; knows nothing about which items are covered or what an incident is
 * worth before the cap bears on it.
 */
final class PayoutCap {

    private static final BigDecimal CAP_MULTIPLE = BigDecimal.valueOf(2);

    private BigDecimal remaining;

    PayoutCap(List<Item> items) {
        this.remaining = InsuranceSum.of(items).multiply(CAP_MULTIPLE);
    }

    /**
     * What is still available to pay out under this policy.
     */
    BigDecimal remaining() {
        return remaining;
    }

    /**
     * Draws the owed amount against the cap, returning what the cap allows and
     * reducing what is left by exactly that much.
     */
    BigDecimal draw(BigDecimal owed) {
        BigDecimal allowed = owed.min(remaining);
        remaining = remaining.subtract(allowed);
        return allowed;
    }
}

File: experiments/runs/2026-09-23_13-26-00_claim-office-example-mapping_exact-ptdd-v1.1-refactor-subagent-cc_opus-5-no-thinking-2/src/payout-cap.ts
import { policyInsuranceSum, type Item } from "./item-pricing.js";

const CAP_MULTIPLE_OF_INSURANCE_SUM = 2;

/**
 * The MHPCO's promise-ceiling rule: the office will ever pay out at most twice
 * a policy's insurance sum, and it treats that ceiling as a finite balance —
 * every payout draws the balance down, and once it is exhausted the office pays
 * nothing more. Establishing the ceiling and drawing on it are the same office
 * decision about how far its promise reaches, so they are stated together here.
 */
export function payoutCapOf(items: Item[]): number {
  return policyInsuranceSum(items) * CAP_MULTIPLE_OF_INSURANCE_SUM;
}

/** What the cap still affords of a desired payout, and what is left afterwards. */
export function drawFromCap(
  remainingCap: number,
  desiredPayout: number,
): { payout: number; remainingCap: number } {
  const affordable = Math.min(desiredPayout, remainingCap);
  return { payout: affordable, remainingCap: remainingCap - affordable };
}
